runtime/vm/ValueTypeHelpers.hpp

```diff
@@ -255,6 +255,82 @@ class VM_ValueTypeHelpers {
 		return rc;
 	}
 
+	static VMINLINE void
+	jitAastore(J9VMThread *currentThread, MM_ObjectAccessBarrierAPI _objectAccessBarrier, j9object_t arrayObject, U_32 index, j9object_t *paramObject, bool fast)
+	{
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+		J9ArrayClass *arrayrefClass = (J9ArrayClass *) J9OBJECT_CLAZZ(currentThread, arrayObject);
+		if (J9_IS_J9CLASS_VALUETYPE(arrayrefClass->componentType)) {
+			if (J9_IS_J9CLASS_FLATTENED(arrayrefClass)) {
+				_objectAccessBarrier.copyObjectFieldsToFlattenedArrayElement(currentThread, arrayrefClass, *paramObject, (J9IndexableObject *) arrayObject, index);
+			} else {
+				_objectAccessBarrier.inlineIndexableObjectStoreObject(currentThread, arrayObject, index, *paramObject);
+			}
+		} else
+#endif /* if defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
+		{
+			_objectAccessBarrier.inlineIndexableObjectStoreObject(currentThread, arrayObject, index, *paramObject);
+		}
+	}
+
+	static VMINLINE j9object_t
+	jitAaload(J9VMThread *currentThread, MM_ObjectAccessBarrierAPI _objectAccessBarrier, MM_ObjectAllocationAPI _objectAllocate, j9object_t arrayObject, U_32 index, bool fast)
+	{
+		j9object_t value = NULL;
+
+		if (fast) {
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+			J9Class *arrayrefClass = J9OBJECT_CLAZZ(currentThread, arrayObject);
+			if (J9_IS_J9CLASS_FLATTENED(arrayrefClass)) {
+				j9object_t newObjectRef = _objectAllocate.inlineAllocateObject(currentThread, ((J9ArrayClass*)arrayrefClass)->leafComponentType, false, false);
+				if (NULL == newObjectRef) {
+					value = newObjectRef;
+					goto done;
+				}
+				_objectAccessBarrier.copyObjectFieldsFromFlattenedArrayElement(currentThread, (J9ArrayClass *) arrayrefClass, newObjectRef, (J9IndexableObject *)arrayObject, index);
+				value = newObjectRef;
+			} else
+#endif /* if defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
+			{
+				value = _objectAccessBarrier.inlineIndexableObjectReadObject(currentThread, arrayObject, index);
+			}
+		} else if (NULL != arrayObject) {
```

Comment:
this else branch is not needed, all the checks are done before calling this function